internal/kubernetes/client.go

```diff
@@ -0,0 +1,172 @@
+	if err := resourceInfo.Get(); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+
+		// If object doesn't exist then create the resource in the given namespace.
+		obj, err := helper.Create(resourceInfo.Namespace, true, resourceInfo.Object)
+		if err != nil {
+			return err
+		}
+		if err := resourceInfo.Refresh(obj, true); err != nil {
+			c.log.Warn(err)
+		}
+	}
+
+	// Patch an OpenAPI resource
+	_, patchedObj, err := patcher.Patch(resourceInfo.Object, modified, resourceInfo.Source, resourceInfo.Namespace,
+		resourceInfo.Name)
+	if err != nil {
+		return err
+	}
+
+	// Refresh, updates the object with another object
+	if err := resourceInfo.Refresh(patchedObj, true); err != nil {
+		c.log.Warn(err)
+	}
+
+	return nil
+}
+
+// Delete resource from cluster using kind, name and namespace.
+func (c *client) deleteResourceByKindAndNameAndNamespace(kind, name, namespace string, do metav1.DeleteOptions) error {
+	gvk, err := c.mapper.KindFor(schema.GroupVersionResource{Resource: kind})
+	if err != nil {
+		return err
+	}
+
+	restMapping, err := c.mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return err
+	}
+
+	restClient, err := newRestClient(*c.config, gvk.GroupVersion())
+	if err != nil {
+		return err
+	}
+	helper := resource.NewHelper(restClient, restMapping)
+
+	// delete resource based on namespaced or non-namespaced.
+	if helper.NamespaceScoped {
+		return c.dynamicClient.Resource(restMapping.Resource).Namespace(namespace).Delete(context.Background(), name, do)
+	}
+	return c.dynamicClient.Resource(restMapping.Resource).Delete(context.Background(), name, do)
+}
+
+func newRestClient(restConfig rest.Config, gv schema.GroupVersion) (rest.Interface, error) {
+	restConfig.ContentConfig = resource.UnstructuredPlusDefaultContentConfig()
+	restConfig.GroupVersion = &gv
+	if len(gv.Group) == 0 {
+		restConfig.APIPath = "/api"
+	} else {
+		restConfig.APIPath = "/apis"
+	}
+
+	return rest.RESTClientFor(&restConfig)
+}
+
+// overlyCautiousIllegalFileCharacters matches characters that *might* not be supported.  Windows is really restrictive,
+// so this is really restrictive
+var overlyCautiousIllegalFileCharacters = regexp.MustCompile(`[^(\w/.)]`)
+
+// computeDiscoverCacheDir takes the parentDir and the host and comes up with a "usually non-colliding" name.
+func computeDiscoverCacheDir(parentDir, host string) string {
+	// strip the optional scheme from host if it's there:
+	schemelesHost := strings.Replace(strings.Replace(host, "https://", "", 1), "http://", "", 1)
+	// now do a simple collapse of non-AZ09 characters.  Collisions are possible but unlikely.  Even if we do collide the problem is short lived
+	safeHost := overlyCautiousIllegalFileCharacters.ReplaceAllString(schemelesHost, "_")
+	return filepath.Join(parentDir, safeHost)
+}
+
+func SetNamespaceIfScoped(namespaceOverride string, u *unstructured.Unstructured, helper *resource.Helper) error {
+	if helper.NamespaceScoped {
```

Comment:
nice, so this is able to look at the ResourceType to determine if it's a namespaced resource?

Reply:
yes

Reply:
Cool!